returnn/tf/layers/rec.py

```diff
@@ -1043,10 +1043,13 @@ def _construct_template(self):
     from pprint import pformat
     from collections import OrderedDict
     from returnn.util.basic import StringIO
-    from returnn.tf.network import NetworkConstructionDependencyLoopException
+    from returnn.tf.network import NetworkConstructionDependencyLoopException, DataNotFound
     # The stack trace is not so interesting for these exceptions.
     skip_stack_trace_exception_types = (
-      NetworkConstructionDependencyLoopException, LayerNotFound)
+      NetworkConstructionDependencyLoopException)
+
+    # These Exceptions always indicate incorrect construction, so fail directly instead of collecting them
+    fail_directly_exception_types = (DataNotFound, LayerNotFound)
```

Comment:
Are we certain that LayerNotFound is really never expected? Because e.g. in `Network.get_layer()` it is raised when a Layer does exist in the network dict, merely is not constructed yet. I think here doing template construction we should be fine, because the layers in the subnet are not expected to be constructed yet and any layer accessed outside should already be constructed. Just wanted to point out that it is not as simple as it sounds.

Reply:
This is why I did not move it, but at least all checks are passing...

Reply:
The construction code should always go through the dynamic `get_layer` (misleading name, `construct_layer` might have been better), so this should not really happen during template construction.

It might have happened at some point in the past. I don't exactly remember anymore. I actually wondered myself why I had it in `skip_stack_trace_exception_types`.

But when all checks are passing now (a lot of things got cleaned up here in the recent past), I think we should really just remove it.

Reply:
So it is okay like it is?

Reply:
From my side, this seems good. Not sure if @patrick-wilken has further comments here.

Reply:
Yes, should be fine.